### PR TITLE
[WIP] Add new interface elements to represent the combat turn order

### LINKED ISSUE #105: Ui: Add a list of portraits that represents the turn order during a combat
From a player point of view, when a combat starts there is now way to know in which order the characters will play. It is a little confusing and I think it is an important feature for a turn-based RPG. 

One solution could be having a small list of portaits, displayed at the top of the screen, that represents the turn order.

Here is a screenshot from [Halcyon 6](http://playhalcyon6.com/press/Halcyon_6_Lightspeed_Edition/images/bandicam%202017-07-11%2008-44-33-647.png) and an other from [Divinity: Original Sin.](http://www.divinityoriginalsin.com/screens/DOS_0040.jpg)

Having such a list will help to:
* determine which character is currently playing
* determine in which order the characters will play during (at least) the current and the next turn

In Halcyon 6, this list is dynamic. That means if a character speed increase or decrease, the list is updated on the fly. Also there are always 12 portraits displayed, twice the number of characters present in the combat.

Let me know what you think about that : )

### PR DESCRIPTION
**Pull request by GresilleSiffle**
-- Work In Progress -- but ready for comments, advices and guidelines : )

In order to close #105, this commit add two new scenes: _TurnOrder_ and _CombatPortrait_.

The _TurnOrder_ scene is designed to be a child of the _CombatInterface_ node in the _CombatArena_ scene. It is initialized at the same time as the other interface nodes, and dynamically display a _CombatPortrait_ node for each battlers, both PC and NPC.

For the moment, the _CombatPortrait_ are just placeholders, and do not display the real battlers portraits.

In order to sum up, **what the code currently does**:
* it displays a list of placeholders to represents the battlers portraits
* when a turn ends, it switches to one battler to another with an animation

**What it does not, but has to**:
* display a real combat portrait for each battler
* when a turn ends, skip portraits of dead battlers and disable them
* wait for the transition to the combat arena to finish before initializing the _TurnOrder_ interface
* hide the _TurnOrder_ interface when the combat ends to properly display the rewards

**What it does not for the moment:, but could** (need your feedback):
* dynamically re-order the combat portraits if a battler joins or leaves during a fight
* dynamically re-order the combat portraits if one of the battler speed changes during a fight

### CODE REVIEW

**Comment by NathanLovato** [comment]
Thanks for your contribution and for the detailed description! In general you only need to explain the gist of your system, especially if it's not too big. 

Reviewers can see everything it does from the code, and if we have any questions we can always ask.

Moving forward you can hit multiple birds in one stone by having the turn order interface react to a signal from TurnQueue: when the TurnQueue node updates it emits the new queue and you update the interface. Queuing it free and re-creating it is fine for a first pass on this system.

File: godot/combat/CombatArena.gd [hunk 1]
```diff
@@ -4,6 +4,7 @@ const BattlerNode = preload("res://combat/battlers/Battler.tscn")
 
 onready var turn_queue : TurnQueue = $TurnQueue
 onready var interface = $CombatInterface
+onready var turn_order = $CombatInterface/TurnOrder
```
[the remark this answers is not in the document]
Comment (GresilleSiffle) [reply godot/combat/CombatArena.gd]: I agree, I will remove that line.

File: godot/combat/CombatArena.gd [hunk 2]
```diff
@@ -94,7 +95,10 @@ func play_turn():
 	var action : CombatAction
 	if battler.stats.health == 0:
 		turn_queue.skip_turn()
-	
+
+	# update the TurnOrder interface by highlighting the next battler portrait
+	turn_order.next()
```
[the remark this answers is not in the document]
Comment (GresilleSiffle) [reply godot/combat/CombatArena.gd]: We can do this with two new signals in _TurnQueue.gd_. We will have something like this:

```gdscript
signal end_of_turn
signal queue_changed

func initialize():
	...
	emit_signal('queue_changed', get_battlers())

func play_turn(action : CombatAction, targets : Array):
	...
	emit_signal('end_of_turn', active_battler)
```

In the future, if we add a method to re-order or re-build the turn queue, the `TurnOrder` interface can be notified by emitting the `queue_changed` signal.

Is that a correct implementation?

File: godot/combat/interface/turn_order/CombatPortrait.gd [hunk 3]
```diff
@@ -0,0 +1,37 @@
+extends Control
+
+class_name CombatPortrait
+
+onready var animation_player : AnimationPlayer = $AnimationPlayer
+
+onready var state = null
+enum states {INITIALIZE, ACTIVE, WAITING, DISABLED}
```
[the remark this answers is not in the document]
Comment (GresilleSiffle) [reply godot/combat/interface/turn_order/CombatPortrait.gd]: In fact, the only one that I really needed was `DISABLED`, to remember when a battler is dead and keep its portrait "disabled". It allows me to know in which state a battler is by just looking at its portrait state, but maybe I won't need states anymore if I use signals to connect the `TurnQueue `and the `TurnOrder `nodes.

File: godot/combat/interface/turn_order/TurnOrder.gd [hunk 4]
```diff
@@ -0,0 +1,38 @@
+extends Control
+
+onready var active_battler_index : int = -1 setget _set_active_battler_index
+
+onready var portraits = $CombatPortraits
+var CombatPortrait = preload("res://combat/interface/turn_order/CombatPortrait.tscn")
+
+func initialize(battlers : Array) -> void:
```
[the remark this answers is not in the document]
Comment (GresilleSiffle) [reply godot/combat/interface/turn_order/TurnOrder.gd]: Undertood. We could call the method `initialize(battlers)` and rebuild the `TurnOrder` list each time a specific signal (`queue_changed` for example) is emitted from the `TurnQueue` node.

File: godot/combat/CombatArena.gd (hunk 2 above)
[the remark this answers is not in the document]
Comment (NathanLovato) [reply godot/combat/CombatArena.gd]: The queue changes at the end of a turn, so you can do with only a `turn_ended` signal I think

File: godot/combat/interface/turn_order/TurnOrder.gd (hunk 4 above)
[the remark this answers is not in the document]
Comment (NathanLovato) [reply godot/combat/interface/turn_order/TurnOrder.gd]: `initialize` suggests a one-time function that you'd call like you would call a constructor
So `initialize` if you need to initialize any property or something, and `rebuild` or something like that to rebuild the menu

**Comment by GresilleSiffle** [comment]
Ok, I've updated the `TurnOrder` to rely only on the `TurnQueue` signals.

Currently, the list of portraits reflects exactly what is in the turn queue. It means that dead battlers don't appear in the list anymore. It may be good for a first iteration. In the future, if we want a more complex system I think we will need to alter the `TurnQueue` behaviour.

... aaand the `TurnOrder` node animations still need some love, more specifically when the `TurnQueue` is updated.

**Comment by NathanLovato** [comment]
I thought dead battlers stayed in the turn queue (for when they're revived) and that the queue just skipped the turn for them. I'll have to check.
Anyway sure we can build this kind of feature in multiple passes 🙂

File: godot/combat/interface/CombatInterface.gd [hunk 5]
```diff
@@ -6,12 +6,14 @@ signal targets_selected(targets)
...
-onready var select_arrow = $SelectArrow
 onready var popup = $PopUpHandler
+onready var select_arrow = $SelectArrow
+onready var turn_order = $TurnOrder
 
 export var buttons_offset : Vector2 = Vector2(-75.0, 0.0)
 
-func initialize(battlers : Array):
+func initialize(battlers : Array, turn_queue : TurnQueue):
+	turn_queue.connect('queue_changed', turn_order, '_on_queue_changed')
```
Comment (NathanLovato) [inline godot/combat/interface/CombatInterface.gd]: Sorry if it wasn't clear from me here. Combat arena should only communicate with the interface's root node, but this node is more of a funnel that will delegate everything to specific UIs. It shouldn't connect to the turn queue itself: you want the turn order list menu to do that instead

**Comment by GresilleSiffle** [comment]
Thanks for the review @NathanLovato, your comments help a lot ; )

I need a little extra time to add the missing docstrings when relevant, and to free the `CombatPortrait` when the battle ends. Also, the last thing to refactor is the connection with signals that is currently done in the `CombatInterface`. After that I think we will be good.

File: godot/combat/interface/CombatInterface.gd (hunk 5 above)
Comment (GresilleSiffle) [reply godot/combat/interface/CombatInterface.gd]: Sorry, maybe I was a little confused, but we are going to make it I'm sure : )

So, the right thing to do is to pass the `TurnQueue` node by argument to the `TurnOrder`? Through the `CombatInterface`?

**Comment by GresilleSiffle** [comment]
I forgot to mention, yes I confirm the dead battlers don't stay in the queue. So it's impossible for the UI to determine if the missing battlers have died, have left, or something else. And it would be a good feature to be able to animate the `CombatPortrait` according to the current state of the battler. The `CombatPortrait` of dead battlers could turn red for example, petrified battler could turn grey, and leaving battlers could just disappear from the `TurnOrder` list.

**Comment by Zireael07** [comment]
@GresilleSiffle: That would be cool but maybe for a latter PR? Better not to have too huge PRs in general, as they take a lot of time to review and finally get in.

**Comment by GresilleSiffle** [comment]
Yes, for a future pull request of course.

**Comment by NathanLovato** [comment]
Okay let's keep dead battlers in the queue and add the ability to revive them later maybe

**Comment by GresilleSiffle** [comment]
I just added the Godette portrait as a placeholder for each combat portrait.

I'm not very good at graphics, also I don't know what is the best way to make a link between a battler scene and its combat portrait. So, currently we have:

turn order screenshot 2

I think the code is ready for a first iteration on this feature, but if we want all portraits (Godette, Robi and Porcupine) to be available before merging I will need the help of a Krita ninja : )

**Comment by GresilleSiffle** [comment]
Sorry, I didn't see the issue #121 so I changed back the title to [WIP] again.

**Comment by GresilleSiffle** [comment]
Rebased on master to integrate the needed `TurnQueue `changes.

**Comment by NathanLovato** [comment]
Thanks for your hard work! I'll merge now

**Pull request merged**